common/lib/new.cc

```diff
@@ -0,0 +1,377 @@
+  }
+
+  void set_new_limit(std::size_t new_limit) {
+    if (new_limit >= (m_heap_limit + MIN_ALLOC_SIZE)) {
+      add_block(
+          // New block address will be directly after the region defined by the
+          // previous limit.
+          &__heap_start + m_heap_limit, new_limit - m_heap_limit);
+
+      m_heap_limit = new_limit;
+    }
+  }
+
+private:
+  // Nodes are added at the end of the list.
+  void add_node(block_node *node) {
+    list_node *node_last = m_head;
+    for (; node_last->m_next; node_last = node_last->m_next) {
+    }
+
+    node_last->m_next = node;
+    node->m_prev = node_last;
+  }
+
+  void add_block(std::byte *address, std::size_t size) {
+    // Assume zero alignment padding.
+    add_node(new (address) block_node{size});
+    m_used += sizeof(block_node);
+    m_free += m_head->m_block.m_size;
+  }
+
+  block_node *m_head = nullptr;
+
+public:
+  // Heap limit is essentially a global var.  It can
+  // be set prior to ctor of blocklist.
+  static constexpr std::size_t HEAP_DEFAULT_LIMIT = 4096;
+  static std::size_t m_heap_limit;
+
+  // accounting.
+  std::size_t m_free = 0;
+  std::size_t m_used = 0;
+};
+
+blocklist &get_free_list() noexcept {
+  static blocklist free_list;
+  return free_list;
+}
+
+void run_new_handler() {
+  const auto newp = std::get_new_handler();
+  if (newp) {
+    newp();
+  }
+  else {
+    // Standard c++ requires throwing bad_alloc here.
+    // TODO: implement throw.
+    // Calling abort() here behaves similarly to a caller that 
+    // does not catch bad_alloc, and therefore has an unhandled exception.
+    abort();
+  }
+}
+} // namespace
+
+std::size_t blocklist::m_heap_limit = SIZE_MAX;
+
+void *operator new(std::size_t size) {
+  const auto block = operator new(size, std::nothrow);
+  if (!block) {
+    run_new_handler();
+  }
+
+  return block;
+}
+
+void *operator new[](std::size_t size) { return operator new(size); }
+
+void *operator new(std::size_t count, const std::nothrow_t &) noexcept {
+  auto &free_list = get_free_list();
+  return free_list.split_block(free_list.find_first_fit(count), count);
+}
+
+void *operator new[](std::size_t count, const std::nothrow_t &) noexcept {
+  return operator new(count, std::nothrow);
+}
+
+void *operator new(std::size_t, void *ptr) { return ptr; }
+
+void *operator new[](std::size_t, void *ptr) { return ptr; }
+
+void operator delete(void *ptr) noexcept {
+  if (!ptr) {
+    return;
+  }
+
+  auto &free_list = get_free_list();
+  free_list.free_block(block::get_block(static_cast<std::byte *>(ptr)));
+  free_list.coallesce_blocks();
+}
+
+void operator delete[](void *ptr) noexcept { operator delete(ptr); }
+
+// helper functions for getting/setting flags in guard_object
+static bool initializerHasRun(std::uint8_t *guard_object) {
+  return (*guard_object != 0);
+}
+
+static void setInitializerHasRun(std::uint8_t *guard_object) {
+  *(guard_object) = 1;
+}
+
+extern "C" {
+int __cxa_guard_acquire(std::uint8_t *guard_object) {
```

Comment:
Mark these functions as weak; that way an OS/application developer can provide their own versions that provide the appropriate guarantees in the presence of custom threading primitives.

Reply:
How practical would it be to change the ABI expected by `clang++ --target=mos`, so that the guard variable doesn't take 8 bytes?  We only utilize 1.  Likewise, no need to return an int (uint8 would be sufficient).

Reply:
No idea; we'd have to look at the implementation in clang and see if it's amenable to change. Clang is surprisingly simple though; I'd be surprised if it's super hard.

Reply:
Created llvm-mos/llvm-mos#137.